src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,388 @@
+                {
+                    return filePath;
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Processes the fetched document model and updates the UI.
+        /// </summary>
+        private async ValueTask UpdateModelAsync(CancellationToken cancellationToken)
+        {
+            if (LatestFetchedDocumentModelIsNotEmpty)
+            {
+                var updatedSymbolTreeSource = LatestFetchedDocumentModel;
+
+                // Switch to UI thread to obtain the search query.
+                await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                var searchQuery = searchBox.Text;
+
+                // Switch to the threadpool to get the lastest text view and filter and sort the model.
+                await TaskScheduler.Default;
+                CodeWindow.GetLastActiveView(out var textView);
+                var activeTextView = EditorAdaptersFactoryService.GetWpfTextView(textView);
+                if (activeTextView is not null)
+                {
+                    if (!string.IsNullOrWhiteSpace(searchQuery))
+                        updatedSymbolTreeSource = DocumentOutlineHelper.Search(updatedSymbolTreeSource, searchQuery);
+
+                    updatedSymbolTreeSource = DocumentOutlineHelper.Sort(updatedSymbolTreeSource, SortOption, LspSnapshot, activeTextView.TextSnapshot);
+
+                    await UpdateSymbolTreeSourceAsync(updatedSymbolTreeSource, cancellationToken).ConfigureAwait(false);
+                    StartHightlightNodeTask();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Switch to the UI thread to update the document symbol tree with the model data
+        /// </summary>
+        private async Task UpdateSymbolTreeSourceAsync(
+            ImmutableArray<DocumentSymbolViewModel> updatedSymbolsTreeItemsSource,
+            CancellationToken cancellationToken)
+        {
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            symbolTree.ItemsSource = updatedSymbolsTreeItemsSource;
+        }
+
+        /// <summary>
+        /// Highlights the symbol node corresponding to the current caret position in the editor.
+        /// </summary>
+        private async ValueTask HightlightNodeAsync(CancellationToken cancellationToken)
+        {
+            await TaskScheduler.Default;
+            if (LatestFetchedDocumentModelIsNotEmpty)
+            {
+                CodeWindow.GetLastActiveView(out var textView);
+                var activeTextView = EditorAdaptersFactoryService.GetWpfTextView(textView);
+                if (activeTextView is not null)
+                {
+                    var documentSymbolModelArray = ((IEnumerable<DocumentSymbolViewModel>)symbolTree.ItemsSource).ToImmutableArray();
```

Comment:
Is the underying source already the ImmutableArray? Or is this making another copy of it?

Reply:
Resolving since this code is outdated